### ISSUE TITLE
Add _cov and _cor methods without keyword args?

### ISSUE THREAD

**Issue by rafaqz** [JuliaLang/julia#32778]
https://github.com/JuliaLang/julia/blob/a8a567e72b669525dc4d3a38af275fe042f44a42/stdlib/Statistics/src/Statistics.jl#L679

The `dims` keywords are used directly in in the `cov` or `cor` methods, meaning it isn't possible to use dispatch on dims. Most (all?) other methods in `Statistics` and `Base` that use `dims` have `_func` methods.

I could add methods for  `_vmean`, `_corm` etc but it's inconsistent, and getting a bit deep in the internal details.

**Comment by JeffBezanson** [comment]
What's the ask here? Remove the `::Int` on dims arguments?

**Comment by rafaqz** [comment]
Yes. Remove `::Integer` from some kwarg dim types, and add wrapper `_cov(X, dims)` etc methods that can use multiple dispatch on dims, where they are missing. `dims::Integer` can go on the underscore method? 

I'll put in a pull request if it's an acceptable idea. For background I'm writing a framework for named dimensions, like an abstract AxisArrays.

When there are underscore methods without `dims::Integer` in the kwargs it's easy to add methods that dispatch on dims and `AbstractArray` : https://github.com/rafaqz/DimensionalData.jl/blob/master/src/abstractdimension.jl#L113

When there is `::Integer` on the dim kwarg, especially with no underscore method, it's not really possible to reuse Base code. Another difficult case is [accumulate](https://github.com/JuliaLang/julia/blob/39e5742557cf56f26e1eccd28ab52d258ae19863/base/accumulate.jl) where adding a single method for `_accumulate!` should be enough for `cumprod`, `accumulate` etc, but actually you need to duplicate half the file.

**Comment by tkf** [comment]
FYI, there is a (sub-)discussion in #16606 about adding a better and more visual way to specify `dims`.  The idea is to create an array-like object which pairs the array and `dims`.  The syntax I used in a proof-of-concept implementation is `sum(along(x, &, &, :))` to mean `sum(x; dims=(1, 2))` https://github.com/JuliaLang/julia/issues/16606#issuecomment-522293582.  Maybe something like `cov(along(x, :time))` could be useful for AxisArray? An advantage is that this reduces number of API functions like `_cov`. (And obvious disadvantage is that discussion like this takes forever to converge...)

**Comment by rafaqz** [comment]
Thanks that's an interesting aproach. In DimensionalData.jl it's `sum(x; dims=Y)`. Specifying the order with `along(x, &, &, :)` is kind of an antipattern with named dimensions, but `sum(along(x, Y))` would be nice. 

`along` might not actually educe api functions for DimensionalData or for AxisArrays. You really want to override the underscore methods anyway to add updated dimensions to the new array - so the named dimension syntax can chain through multiple operations.

In the meantime while `along` is discussed this would make the `sum(x; dims=Y)` etc syntax available to array types that define a `dims()` method with a DimensionalData.jl dep.

**Comment by tkf** [comment]
> but `sum(along(x, Y))` would be nice

Yes, that's what I was actually trying to suggest. (I'm imagining that `Y` is something like `Dim{:x}` or it's singleton instance.)

> `along` might not actually reduce api functions for DimensionalData or for AxisArrays

It was more about reducing API surface for `Statistics`.  Underscore-based approach would mean you need two public APIs (`f` and `_f`) for each function that may take `dims` as an argument.  This approach may be manageable/preferred since the set of functions implemented in `Statistics` and `Base` is not super large and there is probably no future API changes, though.

PS: I just looked at DimensionalData.jl's README. I like xarray functionaries so I appreciate that you are exploring this area!

**Comment by rafaqz** [comment]
`Y` is mostly actually just `Y`/`Lat`/`Lon` (macro generated types) to avoid the ugly `Dim{:x}` syntax, although you can use that too.

With the larger api surface - yes I'm assuming some stability in the `_f` methods when they aren't really a public API. But they probably don't need to be formally... they're more convenient as a developer api than the kwargs methods even without this issue (many of them are used for multiple public methods). So its worth a little extra maintenance if they happen to change.

Also seeing you're interested, there is more high level xarray-like functionality in [GeoData.jl](https://github.com/rafaqz/GeoData.jl). DimensionalData.jl was separated out to useful and lightweight for non-geospatial stuff. Feedback on structural decisions would be great if you have anyway.

**Comment by tkf** [comment]
> some stability in the `_f` methods when they aren't really a public API

I thought you'd want `_f` to be a public (but non-exported) API because otherwise you can't rely on it.  It might get lost during some refactoring in the future unless it's documented that `_f` has to be there.  I don't think "stable but not really public" is a good idea because it's harder for other developers to find out that they can use `_f` to do what they want.

**Comment by rafaqz** [comment]
Bump. 
If I put together a pull request removing `::Integer` from some dims kwargs (mostly in accumulate.jl) would it be accepted? 
Also would adding dispatchable `_methods` for cov, cor and a few others be acceptable?

**Comment by timholy** [comment]
Without seeing the code, it's hard to promise "yes we'll merge it." Best is just to submit it and see what happens. Folks of course have their own tasks so you may not get an instant response, but if you don't get a code-review within a week or so just bump the PR periodically until someone gets around to it.

Obviously, keep in mind the backwards-compatibility promises of the Julia 1.x series. Breakages in purely internal code are largely OK, but not in the public interface.

**Comment by rafaqz** [comment]
Sure. It wont break anything, just type loosenong and some extra intermediate methods without kwargs. I'll make a pull request.
